PP-374: Job with a root-owned script runs even if root_reject_scripts…

nithinj opened the pull request:
<!--- Please review your changes in preview mode -->

<!--- Provide a general summary of your changes in the Title above -->
#### Issue-ID

<!--- Replace XXXX below (2 places) with the actual JIRA id -->
- **[PP-374](https://pbspro.atlassian.net/browse/PP-374)**
#### Problem description
- Job with a root-owned script runs even if root_reject_scripts is set to true in asynchronous communication.
#### Cause / Analysis
- Job with a root-owned script should not run and job comment should get updated with appropriate error message.
#### Solution description
- Dequeue job from the mom's list before sending the reject reply so that job will not run even if a commit message is received.
- Handle root script reject error in post_sendmom and update the ATR_Comment with error.
- In case of asynchronous job server will always reply back scheduler with success and consequently scheduler updates the job comment as job has started running. This block has moved to server so that server will update the comment only after receiving a success from mom.
#### Checklist:

<!--- Use the preview button to see the checkboxes/links properly. -->

<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) _(or)_
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

****_For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).**_**

… is set to true

arungrover commented on src/server/req_quejob.c:
```diff
@@ -1959,6 +1969,7 @@ req_jobscript(struct batch_request *preq)
 	if (fds < 0) {
 		log_err(errno, "req_jobscript", msg_script_open);
+		delete_link(&pj->ji_alljobs);
 		req_reject(PBSE_INTERNAL, 0, preq);
```
Not related to this change... but it should be PBSE_SYSTEM than Internal error

nithinj replied:
This behavior is consistent with rest of the code. We have the same on requests.c:4738, req_quejob.c:2152. What do you recommend?

arungrover replied:
I'd say that you don't change it now but log a bug stating that error numbers returned are wrong.

nithinj replied:
PP-493 has logged for this.

arungrover commented:
Another thing (not related to code)... when you raise a pull request please don't leave spaces around 'x' that you put to mark the check box. It does not really come up as marked that ways.

bhroam commented:
I dislike how you've moved code around.  The scheduler has always owned the comment.  You've moved some of that ownership into the server.  Now the server sets the comment when the job runs.  If a job is in the queued state, the scheduler will set the comment the next time the job is considered to run.  

You've also added one of the "Not Running" messages in the server.  Right now all of those are in one place.  If you want to go looking for them, you look in job_info.c and there's a table of them at the top of the file.  You've moved one of them into the server.

There are many reasons the mom can reject and requeue a job.  You've taken one of them and added a comment to it.  You've made this one different from the rest.  Any time there are caveats to a rule, confusion follows.

Any reason we can't just reject these jobs at submission time?

nithinj commented:
Hi Arun,

The comment is modified only after server receives a reply from mom. But I agree that qstat -f will give inconsistent data to user, there the status will be 'R' but the comment will be the older one. But state is updated to 'R' just before sending the job to mom and substate is updated to PRERUN. User can make sure whether the job has really started running only by checking substate of the job. The job comment will reflect the substate. For more clarity we can add another comment when the job state is updated to 'R' saying - "Job has scheduled to run on $execnode" or "Job has sent for execution on $execnode". But that requires an external design.

nithinj commented:
Hi Bhroam,

The scheduler will always update the comment stating that it has started running in case of asynchronous job irrespective of the fate of the job. This is because scheduler has to proceed without waiting for the reply from server/mom. Should we reassign this task of updating comment to someone else who is more updated about the status of job? (only if job has sent for execution). Please suggest if there is a better alternative to handle this.
“Not Running” message has added to keep consistent with other messages which scheduler is printing. I can remove this and make it in a unique way. I agree that updating error should be in a generic block within post_sendmom since there can be many cases in which mom rejects a job. I think all other cases mentioned in job_info.c are related to scheduling errors.
Can we reject the job while submission? – Server is sending three messages to mom to execute a job in a normal scenario. PBSD_QUEUE, PBSD_Jobscript, PBSD_Commit. The jobscript is send in the second message. If the mom identifies that it is a root script it will reject the job and send the reply back. So server will not send a commit message which will trigger the job in mom(Synchronous). In asynchronous behavior, server sends all three commands without waiting for reply. In the mom side the job is dequeued preventing it from being run but server knows about this rejection only after submission stage.

bhroam commented:
Hi Nithin,
Arun and I had a discussion about this issue.  He convinced me you're doing the correct thing.  Ever since we started using asynchronous run job, the job comments about when the job was run could be erroneous.  If the mom rejected the job, we'd say it ran and it didn't really.  I was also concerned about the scheduler resetting this message in the following cycle.  Arun pointed out that there is no difference with what you are doing and with the pre-13 behavior.  The pre-13 behavior could still change the comment in the following cycle.   What would end up in a lot of the mom-reject situations was a held job with a job running comment.  With your change we will have a more correct error message.  Less confused users means less calls to the admins.  This is always something we strive for.

I found my answer for why the server can't reject these jobs at submission time.  The root_reject_scripts variable is in the mom config.  I was unaware of this config variable.  I thought it was on the server.  Sorry for the confusion.

Thanks for making the change so every mom-reject gets a comment message.  I sign off.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PBSPro/pbspro?pullRequest=176) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/PBSPro/pbspro?pullRequest=176) before we can accept your contribution.<br/>

nithinj commented:
Hi @arungrover ,

I've addressed your comments in recent commits. The old comment will be cleared before sending the job to mom. So we are not giving any room for confusion in qstat -s result. Comment is set once server gets an acknowledgement from mom. Please review and share your feedback.

arungrover commented on src/scheduler/fifo.c:
```diff
@@ -1561,34 +1561,9 @@ run_update_resresv(status *policy, int pbs_sd, server_info *sinfo,
 			update_job_comment(pbs_sd, array, buf);
```
can this comment also move to server?

nithinj replied:
I've kept it in scheduler since it is generic to all the subjobs in the array. We are updating this comment in server after receiving the reply from mom reflecting whether it has started/failed.

arungrover replied:
So in the era of synchronous run_jobs if an array subjob use to fail because of MOM rejects, what use to be the comment set on the array parent?
I think it either stays as "Job Array has Begun" on array parent if the array parent in in 'B' state OR it changes to the job comment being set by scheduler if array parent is not in 'B' state.
If what I mentioned is true (and you may want to test it out :) ) then with Asynchronous run jobs the Array parent will always show "Job Array has Begun" comment even when it's very first subjob fails to run. So, in because of this shifting the comment to server makes sense. Also, moving the job comment for normal jobs to server and keeping for array jobs to scheduler looks inconsistent.

nithinj replied:
The behavior of synchronous jobs are just like you have explained. Even in asynchronous jobs the comment is updated for both normal jobs and array jobs. So if the first job fails the comment will be updated accordingly, replacing the 'begun' comment. For consistency i will move this block to server and place it along with the block where we are resetting the comment for normal jobs.

arungrover commented on src/server/req_runjob.c:
```diff
@@ -1310,6 +1334,16 @@ post_sendmom(struct work_task *pwt)
 				        jobp->ji_qs.ji_destin, pbs_errno);
 			log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, jobp->ji_qs.ji_jobid, log_buffer);
+			snprintf(log_buffer, sizeof(hook_buf),
+					"Not Running: PBS Error: %s", pbse_to_txt(PBSE_MOMREJECT));
```
It's a new job comment. You might want to mention it as an external design change on forum.

bhroam replied:
I'm not necessarily sure this is a new comment.  If the runjob request is rejected, the comment is in the form as above.  So if the mom rejected the request and that error code made it back to the scheduler, the scheduler would print the same thing.  The question is if this would happen without async runjob.  Nithin, could you test this with throughput mode off?  Basically turn off throughput mode and rerun your test.  If the comment is the same with and without throughput mode, you won't need an EDD.

nithinj replied:
They are same. I have attached log in the jira for how it looks like in both the scenarios.

arungrover commented on src/server/req_runjob.c:
```diff
@@ -1310,6 +1334,16 @@ post_sendmom(struct work_task *pwt)
 				        "send of job to %s failed error = %d",
 				        jobp->ji_qs.ji_destin, pbs_errno);
 			log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO, jobp->ji_qs.ji_jobid, log_buffer);
+			snprintf(log_buffer, sizeof(hook_buf),
```
Not that it matters because hook_buf is smaller than log_buffer but sizeof(hook_buf) seems wrong here.

arungrover commented on src/server/req_runjob.c:
```diff
@@ -985,11 +985,15 @@ svr_strtjob2(job *pjob, struct batch_request *preq)
+	/* clearing old comment since we are sending a fresh job */
+	job_attr_def[(int) JOB_ATR_Comment].at_decode(
+			&pjob->ji_wattr[(int) JOB_ATR_Comment], (char *) 0,
+			(char *) 0, "");
```
what will be the job comment from this point to when it is set in complete running or upon mom rejection? Isn't this an external change too?

bhroam replied:
I'm confused on this as well.  Why clear the comment if you are just going to set it again down the line?

arungrover replied:
you are right that we are not introducing a new message/comment but we are momentarily removing the comment from a job which is scheduled to run on a mom and is in substate 41. This wasn't the case earlier, So I think it does impact external behavior. Why don't we set the job comment to "Job run" (as it was earlier) or introduce a new "Job about to run" comment when server gets a run job request. This was we don't have to clear the comment and even mom rejects will show up on the job.

subhasisb replied:
I agree. This is a change in behavior. However, I feel we should not set the comment as "Job run" there, since the job has not actually run yet, and the comment could immediately get replaced by a mom reject message. I like the "job about to run" comment. That is more consistent with the state. But the point is, we need to clear (by setting the above new message) the rejected message from the older cycle, since that could confuse somebody seeing a R/41 with "mom rejected" message.

bhroam replied:
With the server in charge of setting the comment after run, we have the opportunity to be more clear than we had before.  Instead of just saying, "Job is about to run" we can say things like, "Job is staging files" or "Nodes are being provisioned for job".  I dislike just unsetting the comment.  An unset comment has meaning today.  It means the scheduler has not seen the job.

nithinj replied:
I have posted a EDD ( https://pbspro.atlassian.net/wiki/display/PD/EDD+for+PP-374%3A+Job+with+a+root-owned+script+runs+even+if+root_reject_scripts+set+to+true# ) for the proposed changes to update attribute comment with "Job has sent for execution at $execvnode" just before sending job for execution. Please review and provide feedback.

bhroam commented on src/server/req_runjob.c:
```diff
@@ -989,6 +989,17 @@ svr_strtjob2(job *pjob, struct batch_request *preq)
 			(char *) 0, "");
+	if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob &&
+			0 == subjob_index_to_offset(pjob->ji_parentaj,
+					get_index_from_jid(pjob->ji_qs.ji_jobid))) {
```
I'd suggest putting the == 0 at the end.  I suggested this style when the PBS coding standard was being written and it was rejected.  In any case, the reason to use 0 == X is so you don't accidently do X = 0.  In this case, X is a function, so the compiler would complain anyway.

nithinj replied:
I used to follow this style of coding to avoid accidental mistakes. But it will be much more natural to read if the number is on the right side of the operator and this is the standard followed throughout PBS code. I will change this in my next commit.

nithinj commented:
Hi @bhroam and @arungrover ,
I've added a new comment which turned out from the EDD discussion, just before sending the job for execution. Now it will reflect the job status in qstat output. Please review.

bhroam commented:
I like how you moved the common code to the new function.  I am wondering one thing.  You're correctly catching if the job is a subjob when setting the parent array's comment.   What about further down?  We don't want to set the subjob's comment.  Are you checking if a job is not a subjob before trying to set its comment to "Job run at' or "PBS Error" ?

nithinj commented:
Yes Bhroam, it is handled. "Job Run at ..." is set only if job is a subjob. "PBS Error" is set for parent if it is a subjob and for job if it is not. Subjob comment is not being set anywhere in the code.

bhroam commented:
Nithin, thank you for answering my question.  I have one last little thing before I'm happy with the code.  You use and return log_buffer in your new function.  That in itself is fine, but I can foresee a situation where someone takes the output of your new function and then creates a log message using log_buffer.  This would be bad.  It'd be taking the output of log_buffer while copying into it.  I'm not sure how best to fix this.  Creating a static buffer within the function would work.  Maybe a comment in the function header might be enough.  What do you think?

nithinj commented:
Ideally any manipulation to the attribute comment which requires a buffer should reside in the function definition. So I would prefer adding a note in the function header saying not to use the same buffer to copy the return values.

nithinj commented:
Works with me. This has been updated in the function header. Kindly check.

bhroam commented:
Works for me.  I sign off.

nithinj commented:
Hi @arungrover 

You had requested some changes earlier regarding return error codes. I've raised a ticket to do the same: PP-493. Can you re-check and approve the requested changes.

arungrover commented:
Change looks good to me!

arungrover commented:
Please sqaush your commits.

nithinj commented:
I've squashed my commits and made the branch up to date. Please take it further.

arungrover commented:
All looks in place. Merged it in.
